aten/src/ATen/test/cuda_vectorized_test.cu

```diff
@@ -0,0 +1,124 @@
+
+  ASSERT_EQ(can_vectorize_up_to<bool>(ptr), 4);
+  ASSERT_EQ(can_vectorize_up_to<int8_t>(ptr), 4);
+  ASSERT_EQ(can_vectorize_up_to<int16_t>(ptr), 4);
+  ASSERT_EQ(can_vectorize_up_to<int>(ptr), 4);
+  ASSERT_EQ(can_vectorize_up_to<int64_t>(ptr), 4);
+
+  ASSERT_EQ(can_vectorize_up_to<bool>(ptr + 1), 1);
+  ASSERT_EQ(can_vectorize_up_to<int8_t>(ptr + 1), 1);
+
+  ASSERT_EQ(can_vectorize_up_to<bool>(ptr + 2), 2);
+  ASSERT_EQ(can_vectorize_up_to<int8_t>(ptr + 2), 2);
+  ASSERT_EQ(can_vectorize_up_to<int16_t>(ptr + 2), 1);
+
+  ASSERT_EQ(can_vectorize_up_to<bool>(ptr + 4), 4);
+  ASSERT_EQ(can_vectorize_up_to<int8_t>(ptr + 4), 4);
+  ASSERT_EQ(can_vectorize_up_to<int16_t>(ptr + 4), 2);
+  ASSERT_EQ(can_vectorize_up_to<int>(ptr + 4), 1);
+
+  ASSERT_EQ(can_vectorize_up_to<bool>(ptr + 8), 4);
+  ASSERT_EQ(can_vectorize_up_to<int8_t>(ptr + 8), 4);
+  ASSERT_EQ(can_vectorize_up_to<int16_t>(ptr + 8), 4);
+  ASSERT_EQ(can_vectorize_up_to<int>(ptr + 8), 2);
+  ASSERT_EQ(can_vectorize_up_to<int64_t>(ptr + 8), 1);
+}
+
+template <typename scalar_t, int vec_size>
+__global__ void vectorized_copy(scalar_t *dst, scalar_t *src) {
+  using vectorized = policies<64, 4>::vectorized<vec_size>;
+  auto policy = vectorized();
+  scalar_t buf[vectorized::thread_work_size];
+  auto accessor = [&](int index) -> scalar_t & { return buf[index]; };
+  policy.load(accessor, src + 256 * blockIdx.x);
+  policy.store(dst + 256 * blockIdx.x, accessor);
+}
+
+TEST(TestVectorizedMemoryAccess, CopyKernel) {
+  if (!at::cuda::is_available()) return;
+  double *b1 = reinterpret_cast<double *>(buffer1);
+  double *b2 = reinterpret_cast<double *>(buffer2);
+
+  // vec4 copy
+  reset_buffers();
+  cudaDeviceSynchronize();
+  vectorized_copy<double, 4><<<16, 64>>>(b2, b1);
+  cudaDeviceSynchronize();
+  ASSERT_EQ(cudaGetLastError(), cudaSuccess);
+  for (int i = 0; i < 1024; i++) {
+    ASSERT_EQ(buffer1[i].x, buffer2[i].x);
+    ASSERT_EQ(buffer1[i].y, buffer2[i].y);
+    ASSERT_EQ(buffer1[i].z, buffer2[i].z);
+    ASSERT_EQ(buffer1[i].w, buffer2[i].w);
+  }
+
+  // vec2 copy
+  reset_buffers();
+  cudaDeviceSynchronize();
+  vectorized_copy<double, 2><<<16, 64>>>(b2, b1);
+  cudaDeviceSynchronize();
+  ASSERT_EQ(cudaGetLastError(), cudaSuccess);
+  for (int i = 0; i < 1024; i++) {
+    ASSERT_EQ(buffer1[i].x, buffer2[i].x);
+    ASSERT_EQ(buffer1[i].y, buffer2[i].y);
+    ASSERT_EQ(buffer1[i].z, buffer2[i].z);
+    ASSERT_EQ(buffer1[i].w, buffer2[i].w);
+  }
+
+  // vec1 copy
+  reset_buffers();
+  cudaDeviceSynchronize();
+  vectorized_copy<double, 1><<<16, 64>>>(b2, b1);
+  cudaDeviceSynchronize();
+  ASSERT_EQ(cudaGetLastError(), cudaSuccess);
+  for (int i = 0; i < 1024; i++) {
+    ASSERT_EQ(buffer1[i].x, buffer2[i].x);
+    ASSERT_EQ(buffer1[i].y, buffer2[i].y);
+    ASSERT_EQ(buffer1[i].z, buffer2[i].z);
+    ASSERT_EQ(buffer1[i].w, buffer2[i].w);
+  }
+
+  // unaligned
+  for (int i = 0; i < 16; i++) {
+    for (int j = 0; j < 16; j++) {
```

Comment:
isn't running 255 failing tests a bit too much?

Reply:
It only takes 7 ms. So it seems to be acceptable.